pkg/azure/api/validation/validation.go

```diff
@@ -121,6 +126,17 @@ func validateProperties(properties api.AzureVirtualMachineProperties, fldPath *f
 	return allErrs
 }
 
+func validateCloudConfiguration(cloudConfiguration api.CloudConfiguration, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	knownCloudInstances := []string{api.AzurePublicCloudName, api.AzureChinaCloudName, api.AzureGovCloudName}
+
+	if cloudName := cloudConfiguration.Name; !slices.Contains(knownCloudInstances, cloudName) {
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("name"), cloudName, knownCloudInstances))
+	}
+
+	return allErrs
+}
```

Comment:
> How does this work if, let's say, CloudConfiguration is AzureGovernment but the region is China? Shouldn't the CloudConfiguration be validated against the region as well if Region is the source of truth here?

Well sort answer, probably there is no such region in the usgov cloud.

AFAIK, the azure client really forces you to give it the correct configuration explicitly (taken from examples like their [sdk docs](https://github.com/Azure/azure-sdk-for-go/blob/a745cc11844efad7f87bea5dfe21b8a4b5bb81e7/sdk/azcore/cloud/doc.go#L10). It's not really working as in AWS where the region is translated to the correct API endpoints. So the region to "target cloud" that we do here is an optimization of sorts but it's not guaranteed to be the source of truth (but it is for the time being). Though think of it especially in the US (where you have 2+ sovereign cloud instances) that you need to believe that all of these instances will use some sort of prefix to discern the regions compared to the public and there is no real mention in the API of such convention.

Reply:
@rishabh-11 

> Can you also help me understand why this is needed? If I am understanding this correctly, this PR will reduce the visibility to a particular cloud and not beyond it. Won't this already be done by the Location field (name of the region) in the providerSpec?

Continued from the comment above, but are you aware of any contra-indication to my reply above from the API. All the examples show you have to explicitly configure the API instance and not just use the `Location` (which again would be the best approach and how AWS works). So far, we think of it as necessary, rather than a constraint.

Reply:
>So the region to "target cloud" that we do here is an optimization of sorts but it's not guaranteed to be the source of truth (but it is for the time being). 

If that is the case then can we not keep it simple and only get the cloud name from `api.CloudConfiguration` and not derive it from region at all. If it is not specified via `api.CloudConfiguration` then default it to `AzurePublic`. Since this is a new feature it would be actually preferable if `api.CloudConfiguration` is explicitly set for clusters set up in Azure China or Azure Gov and not derived from region in case it misses. WDYT?

Reply:
That was what we thought initially, too. However, we compared that to the way it is handled by other cloud providers (especially AWS) and figured that inferring ot from the region would be a much simpler solution. We also asked our contact at Microsoft and he agreed that as of right now this should work without issues (though he encouraged us to rework this at some point in the future).
We then decided to add this into the provider-extension in addition/as fallback to the explicit CloudConfiguration, as we foresee a need for the CloudConfiguration once support for arbitrary Azure instances is required from us. At that point in time, however, additional changes to the CloudConfiguration are basically guaranteed, as all required endpoints have to be given in that case. Since we do not have any way to test this right now, we decided to do it in a way that does not impede future work and leave it open.

Now, I've removed the CloudConfiguration from this PR again (mostly because I did not read this comment before starting to work on it), but I still somewhat prefer it this way, to be honest. There is much less code, the code that is there is simpler, and once we have both a need and a clear picture for the CloudConfiguration on the provider-extension side in the future we can still re-introduce it.

Reply:
@AndreasBurger - i guess there is some confusion regarding my comment. I was in favor of keeping the `api.CloudConfiguration` and removing the need to derive this information from region. 
As you have also said in your comment above:
```
We also asked our contact at Microsoft and he agreed that as of right now this should work without issues (though he encouraged us to rework this at some point in the future).
```
This is one more reason to not add any technical debt now when introducing new code. Otherwise if you have to change it in future because you can no longer reliably derive this from region then you will have to do the following:

- Introduce `CloudConfiguration` in the API
- Get the value from `CloudConfiguration` if set and if not then continue to derive the value from region.
- Then at some later point of time remove the derive from region logic and depend solely on `CloudConfiguration`

So instead of doing all this i would recommend the following:

- Introduce `api.CloudConfiguration` as you initially proposed in this PR.
- Remove the need to derive this value from region in mcm-provider-azure.
- You can potentially derive this value in gardener-extension-provider-azure from region and when required change it there. This way no more MCM changes will be required.
- Allows adding service endpoints to CloudConfiguration in future if and when required.
- Keeps MCM provide extension code clean.